docs/src/proposals/partitioned-inflationary-rewards-distribution.md

```diff
@@ -0,0 +1,103 @@
+more than 10 seconds. This prolonged computation slows down the network, and can
+cause large number of forks at the epoch boundary, which make the matter even
+worse.
+
+## Proposed Solutions
+
+Instead of computing and reward stake accounts at epoch boundary, we will
+decouple reward computation and reward credit into two phases. At the start of
+the new epoch, a background thread is forked to compute the rewards. This is the
+first phase - reward computation. When reaching block height `N` after the
+epoch, the bank starts the second phase - reward credit, in which the rewards
+are credited to the stake accounts, which will last for `M` blocks.
+
+We call them (a) `calculating interval` (`epoch_start..epoch_start+N`) and (b)
+`credit interval` (`epoch_start+N, epoch_start+N+M`), respectively. And the
+combined interval (`epoch_start..epoch_start+N+M`) is called `rewarding interval`.
+
+For `calculating interval`, `N` is chosen to be sufficiently large so that the
+background computation should have completed and the result of the reward
+computation is available at the end of `calculating interval`. `N` can be fixed
+such as 100 (roughly equivalent to 50 seconds), or chosen as a function of the
+number of stake accounts, `f(num_stake_accounts)`.
+
+In `credit interval`,  the bank will fetch the reward computation results from
+the background thread and start credit the rewards during the next `M` blocks.
+The idea is partition the accounts into `M` partitions. And each block, the bank
+credit `1/M` accounts. The partition is required to be deterministic for the
+current epoch, but must also be random across different epochs. One way to
+achieve these properties is to hash the account's pubkey with some epoch
+dependent values, sort the results, and divide them into `M` bins. The epoch
+dependent value can be the epoch number, total rewards for the epoch, the leader
+pubkey for the epoch block, etc. `M` can be choses based on 50K account per
+block, which equal to `ceil(num_stake_accounts/50,000)`.
+
+In order to avoid putting extra burden of computing and credit the stake reward
+for blocks produced during the `rewarding interval`, we can reduce the compute
+budget limits on those blocks in `rewarding interval`, and reserve some computing
+and read/write capacity to perform stake rewarding.
+
+### Challenges
+
+1. stake accounts reads/writes during the `rewarding interval`
+
+`epoch_start..epoch_start+N+M` Because of the delayed credit of the rewards,
+Reads to those stake accounts may not will return the value that the user are
+expecting (viz. not include the recent epoch stake rewards). Writes to those
+stake accounts will be lost once the reward are credited on block
+`epoch_start+N+M`. We will need to modify the runtime to restrict read/writes to
+stake accounts during the `rewarding interval`. Any transactions, which involves
+stake accounts, will result in a new execution error, i.e. "stake rewards
```

Comment:
do we need to restrict during the `calculating interval` as well or are we saving off the necessary data at epoch boundaries?

Reply:
Yeah. Restriction is enforced for the whole `rewarding interval`.

Reply:
Yeah. The plan is to take a ref of the stake cache.